sdk/src/main/java/com/mapbox/maps/MapController.kt

```diff
@@ -226,61 +243,72 @@ internal class MapController : MapPluginProviderDelegate, MapControllable {
   // Plugin API
   //
 
-  override fun <T> getPlugin(clazz: Class<T>): T? = pluginRegistry.getPlugin(clazz)
-
-  override fun <T> getPlugin(className: String): T? {
-    @Suppress("UNCHECKED_CAST")
-    return getPlugin(
-      Class.forName(
-        className
-      )
-    ) as T
+  override fun getPlugin(id: String): MapPlugin? {
+    return pluginRegistry.getPlugin(id)
   }
 
-  fun <T> createPlugin(
+  fun createPlugin(
     mapView: MapView?,
-    clazz: Class<T>,
-    vararg constructorArguments: Pair<Class<*>, Any>
-  ): T? = pluginRegistry.createPlugin(mapView, mapInitOptions, clazz, *constructorArguments)
+    plugin: Plugin
+  ) = pluginRegistry.createPlugin(mapView, mapInitOptions, plugin)
 
   fun initializePlugins(
     options: MapInitOptions,
     mapView: MapView? = null,
   ) {
-    for (pluginName in options.plugins) {
+    for (plugin in options.plugins) {
       try {
-        val pluginClass = Class.forName(pluginName)
-        val plugin = if (pluginName == PLUGIN_GESTURE_CLASS_NAME) {
-          val attrs = options.attrs
-          if (attrs != null) {
-            createPlugin(
-              mapView,
-              pluginClass,
-              Pair(Context::class.java, options.context),
-              Pair(AttributeSet::class.java, attrs),
-              Pair(Float::class.java, options.mapOptions.pixelRatio)
-            )
-          } else {
-            createPlugin(
-              mapView,
-              pluginClass,
-              Pair(Context::class.java, options.context),
-              Pair(Float::class.java, options.mapOptions.pixelRatio)
-            )
+        val pluginObject = when (plugin.id) {
+          MAPBOX_GESTURES_PLUGIN_ID -> {
+            val attrs = options.attrs
+            if (attrs != null) {
+              GesturesPluginImpl(options.context, attrs, options.mapOptions.pixelRatio)
+            } else {
+              GesturesPluginImpl(options.context, options.mapOptions.pixelRatio)
+            }
```

Comment:
Now thinking of this part of the initialisation of GesturesPlugin, why does it need to be a separate constructor for context, attrs and pixel ratio? As far as I see, the GesturesPlugin extends ContextBinder
```
fun interface ContextBinder {
  /**
   * Bind the ViewPlugin with current map context. This will create a View that
   * will be added to the MapView.
   *
   * @param context The hosting context
   * @param attrs parent attributes
   * @param pixelRatio the pixel ratio of the device
   * @return View that will be added to the MapView
   */
  fun bind(context: Context, attrs: AttributeSet?, pixelRatio: Float)
}
```
So the GesturesPlugin can be initialised properly when the plugin is created through the MapPluginRegistry.

Reply:
@pengdev if we think even deeper there is basically no sense in creating Mapbox plugins each time from scratch, we could create them just once and re-use in all upcoming `MapView` (we just need to make sure we `cleanup` properly in `MapView.onDestroy`.

My idea was to basically retain the existing logic and not overcomplicate this PR.
WDYT of leaving it as is here and creating a follow-up task?

Reply:
> we could create them just once and re-use in all upcoming MapView (we just need to make sure we cleanup properly in MapView.onDestroy).

I don't think that's valid, different MapView have different states, we can not simply use the same instance for all the maps, let's say two MapViews in the activity or the multi display use cases.

> My idea was to basically retain the existing logic and not overcomplicate this PR.
> WDYT of leaving it as is here and creating a follow-up task?

Agree with this part, this can be tracked in a separate ticket and addressed as a follow-up.

Reply:
Good catch about multiple map views - completely forgot about it!

Reply:
https://github.com/mapbox/mapbox-maps-android/issues/522